### DISCUSSION TITLE
Specific prices priority over catalog rules - a logic approach

### DISCUSSION THREAD

**Discussion by Digitalfastmind** [PrestaShop/PrestaShop#33440] [Ideas]
**Is your feature request related to a problem?**

Yes, on P.s. 1.7.5
When setting up a catalog rule if the product has already a specific price it gets ignored.
This maybe right if the catalog rule is lower than the actual product specific price, but it's not if the discount can be higher

**Describe the solution you'd like**

We used this workaround in 1.6 and i think it made sense:
https://www.prestashop.com/forums/topic/438626-set-priority-on-product-discount/

**Describe alternatives you've considered**

Hard to say, as if there is for example a black friday, the catalog price rules are best but this way as most of the items can already have a small discount, they will be ignored.

**Comment by khouloudbelguith** [comment]
Hi @freeman78ita,

Yes, When we use specific prices, the catalog price rule is ignored( lower or higher than the specific price).
ping @marionf, @colinegin what do you think of his suggestion.

Thanks!

**Comment by ghost** [comment]
Yes. Catalog price rules should be on top of the priority , then will be useful for merchants.

Case
1) Got dozens of discounted products in category (20 euro, 10%, 15%)
2) Want to make sale offer for this category by the catalog price rule like 20% off
3) Only non distounted products will be covered by the catalog price rule, and rest will have previous special prices  (20 euro, 10%, 15%)
4) To make offer work i need to remove all specific prices from dozens of products...
5) After sale is over i need to add previous removed specific prices as before...

**Comment by colinegin** [comment]
@matks do you think we could add a new option in the catalog price rule page ? 
The option would define if the catalog price rule has or does not have priority over specific prices that may exist for some products impacted by the catalog price rule.

Do you think it could it be done with this issue #10548 ?

**Comment by ghost** [comment]
That would be great!

**Comment by khouloudbelguith** [comment]
Hi @colinegin,

Any news about the fix of this issue? because @PrestaShark said
> We have now an very hot months in e-commerce world. Black Friday, Cyber Monday and Christmas

Thanks!

**Comment by ghost** [comment]
As mentioned above :

"**Example**
I have 500 products in catalog, 2/3 of them have set specific prices from 15 to 40% discounts.
I want to set catalog rule to cover whole site with 50% discount.
Catalog rule will work only with 1/3 of non discounted products because.

CATALOG RULES doesn't work OVER product SPECIFIC PRICE! ;(
Without this the Catalog Rules don't make any sense... "

THANKS!

**Comment by ghost** [comment]
Bump

**Comment by burakgit** [comment]
so @freeman78ita , 
what happened to fix this ? 

@khouloudbelguith it is very bad, 
if i create a new customer there should be a option which enables, disables specific prices, if it cant, catalog prices must ovverride the specific price.

**Comment by Digitalfastmind** [comment]
Nothing......it's a very basic thing. Now for example under covid we would love to set up let's say a category "t-shirt" with 30% off that override specific prices and we can't set it.
I hope they will release this soon. I'm going to start playing with the code myself if i have time and contribute if i can.

**Comment by burakgit** [comment]
There is a paid module, but my country is not supporting Paypal. So i am waiting.. it is so important to the " Disable/enable specific prices for selected groups.

**Comment by Digitalfastmind** [comment]
which module? I think this should be not managed by a plugin but as a native feature.

**Comment by andrevich76** [comment]
hi, any news on the fixing?

**Comment by ghost** [comment]
Guys.... its over two years and nothing was happend here https://github.com/PrestaShop/PrestaShop/issues/16295 

Catalog price priority shoud be over (on top) of specific prices of single products because they are not usable right now.

**Example**
I have 500 products in catalog, 2/3 of them have set specific prices from 15 to 40% discounts.
I want to set catalog rule to cover whole site with 50% discount
Catalog rule will work only with 1/3 of non discounted products because CATALOG RULES doesn't work OVER product SPECIFIC PRICES :(

**Comment by matks** [comment]
Hi @PrestaShark I have read a very accurate answer 2 days ago on Twitter: https://twitter.com/adamwathan/status/1381275398646927362

_"One of the most stressful parts of product development (especially open source) is when people are frustrated you haven’t made the improvement they want and automatically attribute it to complacency or ignoring the community instead of just being completely buried in work to do._

_If something important takes two years to get worked on, it’s usually because there are two years worth of other fires to put out and work that’s necessary for pure survival. Things taking a long time is more stressful for me than it is for you, I promise. We’re on the same side."_

It's Adam Wathan, founder of Tailwind, another popular opensource project, who said that.

Unfortunately it's exactly the same for us. We're buried in work to do. This is why things take years.

If you really need something to go forward on the project and we cannot do it because we have so much work to do, the best is to _invest_ into having it done. You can try doing it on your own. The most important thing is to share that fix with everyone by submitting a Pull Request -- that's what the open source spirit is all about.

**Comment by lucastes** [comment]
Temporary solution - works for me (PS 1.7.6):
1. copy file `/classes/SpecificPrice.php`
2. edit the file, go to line 556 and change ",**` id_specific_price_rule` ASC" -> ` id_specific_price_rule` DESC**

**BEFORE**
$ query. = 'ORDER BY `id_product_attribute` DESC,` id_cart` DESC, `from_quantity` DESC,` id_specific_price_rule` **ASC**, `score` DESC,` to` DESC, `from` DESC';


**AFTER**
$ query. = 'ORDER BY `id_product_attribute` DESC,` id_cart` DESC, `from_quantity` DESC,` id_specific_price_rule` **DESC**, `score` DESC,` to` DESC, `from` DESC';

3. save the changes and move the file to the directory
`/override/classes/SpecificPrice.php`

4. delete the cache - the entire directory `/var/cache/prod` 

Now CATALOG RULES are applied before the SPECIFIC PRICES

**Comment by TouchCab** [comment]
> Temporary solution - works for me (PS 1.7.6):
> 
> 1. copy file `/classes/SpecificPrice.php`
> 2. edit the file, go to line 556 and change ",**` id_specific_price_rule` ASC" -> ` id_specific_price_rule` DESC**
> 
> **BEFORE**
> $ query. = 'ORDER BY `id_product_attribute` DESC,` id_cart` DESC, `from_quantity` DESC,` id_specific_price_rule` **ASC**, `score` DESC,` to` DESC, `from` DESC';
> 
> **AFTER**
> $ query. = 'ORDER BY `id_product_attribute` DESC,` id_cart` DESC, `from_quantity` DESC,` id_specific_price_rule` **DESC**, `score` DESC,` to` DESC, `from` DESC';
> 
> 1. save the changes and move the file to the directory
>    `/override/classes/SpecificPrice.php`
> 2. delete the cache - the entire directory `/var/cache/prod`
> 
> Now CATALOG RULES are applied before the SPECIFIC PRICES

I believe this only works if you want the latest rule to apply, not the largest available discount...

From the thread here (PS 1.6) : https://www.prestashop.com/forums/topic/438626-set-priority-on-product-discount/

It appears that this should be the solution but I am not sure if it works for 1.7 or what needs changed to make it work?

This would be the override (for 1.6):
<?php

class SpecificPrice extends SpecificPriceCore {

    public static function getSpecificPrice($id_product, $id_shop, $id_currency, $id_country, $id_group, $quantity, $id_product_attribute = null, $id_customer = 0, $id_cart = 0, $real_quantity = 0) {
        if (!SpecificPrice::isFeatureActive()) {
            return array();
        }

        /*
         * * The date is not taken into account for the cache, but this is for the better because it keeps the consistency for the whole script.
         * * The price must not change between the top and the bottom of the page
         */

        $key = ((int)$id_product . '-' . (int)$id_shop . '-' . (int)$id_currency . '-' . (int)$id_country . '-' . (int)$id_group . '-' . (int)$quantity . '-' . (int)$id_product_attribute . '-' . (int)$id_cart . '-' . (int)$id_customer . '-' . (int)$real_quantity);
        if (!array_key_exists($key, SpecificPrice::$_specificPriceCache)) {
            $now = date('Y-m-d H:i:s');
            $query = '
            SELECT *, ' . SpecificPrice::_getScoreQuery($id_product, $id_shop, $id_currency, $id_country, $id_group, $id_customer) . '
                FROM `' . _DB_PREFIX_ . 'specific_price`
                WHERE `id_product` IN (0, ' . (int)$id_product . ')
                AND `id_product_attribute` IN (0, ' . (int)$id_product_attribute . ')
                AND `id_shop` IN (0, ' . (int)$id_shop . ')
                AND `id_currency` IN (0, ' . (int)$id_currency . ')
                AND `id_country` IN (0, ' . (int)$id_country . ')
                AND `id_group` IN (0, ' . (int)$id_group . ')
                AND `id_customer` IN (0, ' . (int)$id_customer . ')
                AND
                (
                    (`from` = \'0000-00-00 00:00:00\' OR \'' . $now . '\' >= `from`)
                    AND
                    (`to` = \'0000-00-00 00:00:00\' OR \'' . $now . '\' <= `to`)
                )
                AND id_cart IN (0, ' . (int)$id_cart . ') ';

            if ($real_quantity != 0 && !Configuration::get('PS_QTY_DISCOUNT_ON_COMBINATION')) {
                $query .= ' AND IF(`from_quantity` > 1, `from_quantity`, 0) <= IF(id_product_attribute=0,' . (int)$quantity . ' ,' . (int)$real_quantity . ')';
            }
            else {
                $qty_to_use = $id_cart ? (int)$quantity : (int)$real_quantity;
                $query .= 'AND `from_quantity` <= ' . max(1, $qty_to_use);
            }

 

             /* order by reduction_type and reduction */
            $query .= ' ORDER BY `reduction_type` DESC, `reduction` DESC, `id_product_attribute` DESC, `from_quantity` DESC, `id_specific_price_rule` ASC, `score` DESC';

            SpecificPrice::$_specificPriceCache[$key] = Db::getInstance(_PS_USE_SQL_SLAVE_)->getRow($query);
        }
        return SpecificPrice::$_specificPriceCache[$key];
    }

}

**Comment by mrkalchemy** [comment]
This bug/feature was driving us insane. We couldn't workout why it was ignoring our products custom pricing and using the default price when applying the discount from Catalog Price Rules. We didn't realise that the rules worked like that, as it is not mentioned in the documentation.

We spent a few hours trying to locate the cause of the issue, but our limited knowledge of PrestaShop's internals was hindering our efforts.

Would it be a good idea to add a note to the documentation explaining that Catalog Price Rules work on the default product price only? This might help to reduce confusion for new users in the future.

**Comment by kpodemski** [comment]
Hello everyone. Following our new approach to feature requests, I converted this Issue to the GitHub Discussion so that the community can discuss the best possible ways to handle this problem and maybe, I hope, organize together to work towards fixing it in the core. If you are motivated to do so, we will be happy to help.

**Reply by TouchCab** [reply]
For me, it's simple. Let the user decide! In some cases, if you're very well organised, and you only have a few products, it's easy to remove specific discounts, or to order them in a way that suits you.
In other cases, such as ours, with tens of thousands of products, with specific and catalogue rules, with customers who enjoy larger discounts to others, with flash sales depending on category - it's too hard to do right now. There's not even a way to delete specific discounts on a CSV import - so if i re-import the catalogue via CSV, with the specific discount blank in amount and percent, it keeps it's original discount.

So, why not just let the merchant decide what works for them.

•  Apply discounts in the order they are applied in the list - specific first
•  Apply category discounts over specific discounts in order they are in the list
•  Apply the newest discount first (actually makes more sense as a default to the current default - you're not going to add a discount because you want it ignored!)
•  Apply the largest discount available to the customer regardless of any order in the list, specific or catalogue.

Also, while we're working on this - why not add the ability to add the catalogue price rule as an "additional" discount to the specific one. i.e. if i have products with 10 to 30% off with specific prices, why not allow the merchant to give an additional % discount if these products are in a specific category, a specific brand etc. - Rather than one rule over-riding the other.

I am afraid i am not able to offer the solution to do this, as i am a merchant with "little" coding experience, but I am a merchant with a good deal of experience in using PS!

**Comment by redtyper** [comment]
Hello, 
Do you have a solution but the opposite? The price catalog rule will not be active when the product has a specific price?

**Comment by Digitalfastmind** [comment]
Abandoned presta as dev

Il mer 8 mag 2024, 21:45 RedTyper ***@***.***> ha scritto:

**Comment by redtyper** [comment]
Maybe someone has an idea how to solve this?

**Comment by Digitalfastmind** [comment]
Nope. switched to Shopify and never looked back


Il mer 8 mag 2024, 22:11 RedTyper ***@***.***> ha scritto:

**Comment by Digitalfastmind** [comment]
You can use plugins and pay far less. As usual it all depends on use cases
and needs. Eventually switching to woo would be even a possible solution.
I'm done with presta, all the best of luck with it.

Il gio 9 mag 2024, 09:27 Chillcode ***@***.***> ha scritto:
